<a href="https://colab.research.google.com/github/briandankowski/GuidedCapstone/blob/master/Final_GuidedCapstone_Step4_and_Step5HL.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [ ]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [ ]:
os.listdir()

['.config', 'figures', 'data', 'models', 'step3_output.csv', 'sample_data']

In [ ]:
!mkdir data
!mkdir figures
!mkdir models
from google.colab import files
files.upload()

mkdir: cannot create directory ‘data’: File exists
mkdir: cannot create directory ‘figures’: File exists
mkdir: cannot create directory ‘models’: File exists


Saving step3_output.csv to step3_output (1).csv


{'step3_output.csv': b"Name,state,summit_elev,vertical_drop,fastQuads,quad,triple,double,surface,total_chairs,Runs,TerrainParks,LongestRun_mi,SkiableTerrain_ac,Snow Making_ac,daysOpenLastYear,yearsOpen,averageSnowfall,AdultWeekday,AdultWeekend,projectedDaysOpen,NightSkiing_ac,clusters\nAlyeska Resort,Alaska,3939,2500,2,2,0,0,2,7,76.0,2.0,1.0,1610.0,113.0,150.0,60.0,669.0,65.0,85.0,150.0,550.0,0\nEaglecrest Ski Area,Alaska,2600,1540,0,0,0,4,0,4,36.0,1.0,2.0,640.0,60.0,45.0,44.0,350.0,47.0,53.0,90.0,0.0,0\nHilltop Ski Area,Alaska,2090,294,0,0,1,0,2,3,13.0,1.0,1.0,30.0,30.0,150.0,36.0,69.0,30.0,34.0,152.0,30.0,0\nArizona Snowbowl,Arizona,11500,2300,0,2,2,1,2,8,55.0,4.0,2.0,777.0,104.0,122.0,81.0,260.0,89.0,89.0,122.0,0.0,1\nSunrise Park Resort,Arizona,11100,1800,1,2,3,1,0,7,65.0,2.0,1.2,800.0,80.0,115.0,49.0,250.0,74.0,78.0,104.0,80.0,1\nYosemite Ski & Snowboard Area,California,7800,600,0,0,1,3,1,5,10.0,2.0,0.4,88.0,174.87323943661968,110.0,84.0,300.0,47.0,47.0,107.0,0.0,1\nBear Mountain,

In [ ]:
os.listdir()

['.config',
 'figures',
 'data',
 'models',
 'step3_output (1).csv',
 'step3_output.csv',
 'sample_data']

In [ ]:
!rm 'step3_output (1).csv'

In [ ]:
os.listdir()

['.config', 'figures', 'data', 'models', 'step3_output.csv', 'sample_data']

In [ ]:
os.getcwd()

'/content'

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [ ]:
df = pd.read_csv('step3_output.csv', index_col= None)
df.describe().T

count         mean  ...          75%      max
summit_elev        330.0  4591.818182  ...  7806.000000  13487.0
vertical_drop      330.0  1215.427273  ...  1800.000000   4425.0
fastQuads          330.0     1.018182  ...     1.000000     15.0
quad               330.0     0.933333  ...     1.000000      8.0
triple             330.0     1.500000  ...     2.000000      8.0
double             330.0     1.833333  ...     3.000000     14.0
surface            330.0     2.621212  ...     3.000000     15.0
total_chairs       330.0     8.266667  ...    10.000000     41.0
Runs               330.0    48.214724  ...    60.000000    341.0
TerrainParks       330.0     2.820789  ...     3.000000     14.0
LongestRun_mi      330.0     1.433231  ...     2.000000      6.0
SkiableTerrain_ac  330.0   739.801223  ...   739.801223  26819.0
Snow Making_ac     330.0   174.873239  ...   192.000000   3379.0
daysOpenLastYear   330.0   115.103943  ...   130.000000    305.0
yearsOpen          330.0    63.656535  ...    68.750000   2019.0
averageSnowfall    330.0   185.316456  ...   278.750000    669.0
AdultWeekday       330.0    57.916957  ...    67.000000    179.0
AdultWeekend       330.0    64.166810  ...    72.000000    179.0
projectedDaysOpen  330.0   120.053004  ...   136.000000    305.0
NightSkiing_ac     330.0    56.890909  ...    83.750000    650.0
clusters           330.0     0.357576  ...     1.000000      2.0

[21 rows x 8 columns]

In [ ]:
df.head()

Name    state  ...  NightSkiing_ac  clusters
0       Alyeska Resort   Alaska  ...           550.0         0
1  Eaglecrest Ski Area   Alaska  ...             0.0         0
2     Hilltop Ski Area   Alaska  ...            30.0         0
3     Arizona Snowbowl  Arizona  ...             0.0         1
4  Sunrise Park Resort  Arizona  ...            80.0         1

[5 rows x 23 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [ ]:
# select object type columns
df = pd.concat([df.drop('state', axis=1), pd.get_dummies(df['state'])], axis=1)
df.head()

Name  summit_elev  ...  Wisconsin  Wyoming
0       Alyeska Resort         3939  ...          0        0
1  Eaglecrest Ski Area         2600  ...          0        0
2     Hilltop Ski Area         2090  ...          0        0
3     Arizona Snowbowl        11500  ...          0        0
4  Sunrise Park Resort        11100  ...          0        0

[5 rows x 57 columns]

## Standardize the magnitude of numeric features
**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [ ]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called x,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
x = df.drop(['Name','AdultWeekend'], axis=1)
print(x)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter x 
scaler = preprocessing.StandardScaler().fit(x)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter x 
x_scaled=scaler.transform(x)

     summit_elev  vertical_drop  fastQuads  ...  West Virginia  Wisconsin  Wyoming
0           3939           2500          2  ...              0          0        0
1           2600           1540          0  ...              0          0        0
2           2090            294          0  ...              0          0        0
3          11500           2300          0  ...              0          0        0
4          11100           1800          1  ...              0          0        0
..           ...            ...        ...  ...            ...        ...      ...
325         9500           1000          0  ...              0          0        1
326         7428            810          0  ...              0          0        1
327         7808           1571          0  ...              0          0        1
328         9663            990          0  ...              0          0        1
329         9500           1100          0  ...              0          0        1

[33

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [ ]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()
# Call the train_test_split() function with the first two parameters set to x_scaled and y 
# Declare four variables, s_train, s_test, y_train and y_test separated by commas 
s_train, s_test, y_train, y_test = train_test_split(x_scaled, y, test_size=0.25, random_state=1)

In [ ]:
y

array([ 85.        ,  53.        ,  34.        ,  89.        ,
        78.        ,  47.        ,  64.16681004,  64.16681004,
        64.16681004,  78.        ,  75.        ,  64.16681004,
        64.16681004,  64.16681004, 159.        ,  59.        ,
        84.        ,  69.        ,  64.16681004,  64.16681004,
        83.        ,  64.16681004,  79.        ,  50.        ,
       125.        ,  69.        ,  85.        , 179.        ,
        64.16681004,  64.16681004, 158.        ,  64.16681004,
        89.        ,  64.16681004,  25.        ,  64.16681004,
        79.        ,  89.        ,  71.        ,  79.        ,
        56.        ,  84.        ,  64.16681004,  65.        ,
       139.        ,  64.16681004,  64.16681004,  72.        ,
        65.        ,  60.        ,  55.        ,  62.        ,
        42.        ,  64.        ,  70.        ,  42.        ,
        47.        ,  32.        ,  47.        ,  64.16681004,
        81.        ,  62.        ,  43.        ,  64.16

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

# **Model 1**

In [ ]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(s_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [ ]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(s_test)

In [ ]:
y_pred

array([ 57.92905836,  91.47593336,  79.14780836,  76.33530836,
        68.92905836,  77.86655836,  43.06968336,  47.75718336,
        91.58530836,  85.94468336,  51.53843336,  76.36655836,
        50.05405836,  70.44468336,  63.24155836,  91.77280836,
        42.88218336,  73.74155836,  63.19468336,  41.71030836,
       114.22593336,  37.77280836,  54.11655836,  49.16343336,
        48.96030836,  60.58530836,  77.29429273,  67.99155836,
        62.39780836,  66.52280836,  34.71030836,  58.46030836,
        80.64780836,  57.91343336,  75.91343336,  49.52280836,
        24.11655836,  29.64780836,  60.02280836,  38.42905836,
        49.85093336,  61.41929273,  87.92905836,  71.35093336,
        71.69468336,  59.28843336,  60.80405836, 166.64780836,
        35.19468336,  60.04429273,  66.17905836,  61.63218336,
        51.46030836,  43.91343336,  60.29429273,  42.89780836,
        85.41929273,  36.63218336, 120.13218336, 120.75718336,
        56.24155836,  52.73179273,  41.46030836,  81.11

## Review Model Outcomes — Iterate over additional models as needed

In [ ]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
from sklearn.metrics import explained_variance_score
from sklearn.metrics import mean_absolute_error

In [ ]:
explained_variance_score(y_true=y_test, y_pred=y_pred)

0.9416576172466851

In [ ]:
mean_absolute_error(y_true=y_test, y_pred=y_pred)

4.84926757962234

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [ ]:
lm.intercept_

64.10679273198181

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [ ]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coef_data = []
coef_data = pd.DataFrame(abs(lm.coef_), x.columns, columns=['Coefficient'])
print(coef_data)


                    Coefficient
summit_elev        1.765039e+00
vertical_drop      1.008410e+00
fastQuads          8.335971e-01
quad               1.562365e+00
triple             2.013590e+00
double             1.289321e+00
surface            1.934466e-01
total_chairs       7.694186e-01
Runs               1.445123e+00
TerrainParks       1.781094e-01
LongestRun_mi      2.399679e-01
SkiableTerrain_ac  1.274107e-02
Snow Making_ac     2.236279e-01
daysOpenLastYear   5.430303e-01
yearsOpen          8.975270e-03
averageSnowfall    1.034181e-01
AdultWeekday       1.960756e+01
projectedDaysOpen  1.043239e-01
NightSkiing_ac     5.341784e-01
clusters           6.192308e-01
Alaska             1.494926e+13
Arizona            1.222467e+13
California         3.844802e+13
Colorado           3.928907e+13
Connecticut        1.924030e+13
Idaho              2.948420e+13
Illinois           1.723550e+13
Indiana            1.222467e+13
Iowa               1.494926e+13
Maine              2.565422e+13
Maryland

In [ ]:
coef_data.sort_values(['Coefficient'], ascending=False)

Coefficient
New York           4.725201e+13
Michigan           4.459307e+13
Colorado           3.928907e+13
California         3.844802e+13
Pennsylvania       3.668953e+13
New Hampshire      3.383062e+13
Wisconsin          3.383062e+13
Vermont            3.280847e+13
Minnesota          3.174627e+13
Utah               3.063983e+13
Idaho              2.948420e+13
Montana            2.948420e+13
Massachusetts      2.827332e+13
Washington         2.699977e+13
Oregon             2.699977e+13
Maine              2.565422e+13
New Mexico         2.565422e+13
Wyoming            2.422468e+13
North Carolina     2.104424e+13
Connecticut        1.924030e+13
Ohio               1.924030e+13
West Virginia      1.723550e+13
Virginia           1.723550e+13
Illinois           1.723550e+13
Nevada             1.723550e+13
Iowa               1.494926e+13
Alaska             1.494926e+13
New Jersey         1.222467e+13
Missouri           1.222467e+13
Indiana            1.222467e+13
South Dakota       1.222467e+13
Arizona            1.222467e+13
Tennessee          8.657312e+12
Maryland           8.657312e+12
Rhode Island       8.657312e+12
AdultWeekday       1.960756e+01
triple             2.013590e+00
summit_elev        1.765039e+00
quad               1.562365e+00
Runs               1.445123e+00
double             1.289321e+00
vertical_drop      1.008410e+00
fastQuads          8.335971e-01
total_chairs       7.694186e-01
clusters           6.192308e-01
daysOpenLastYear   5.430303e-01
NightSkiing_ac     5.341784e-01
LongestRun_mi      2.399679e-01
Snow Making_ac     2.236279e-01
surface            1.934466e-01
TerrainParks       1.781094e-01
projectedDaysOpen  1.043239e-01
averageSnowfall    1.034181e-01
SkiableTerrain_ac  1.274107e-02
yearsOpen          8.975270e-03

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

# Model 2 

In [ ]:
df2 = pd.read_csv('step3_output.csv', index_col= None)

In [ ]:
df2.head()

Name    state  ...  NightSkiing_ac  clusters
0       Alyeska Resort   Alaska  ...           550.0         0
1  Eaglecrest Ski Area   Alaska  ...             0.0         0
2     Hilltop Ski Area   Alaska  ...            30.0         0
3     Arizona Snowbowl  Arizona  ...             0.0         1
4  Sunrise Park Resort  Arizona  ...            80.0         1

[5 rows x 23 columns]

In [ ]:
# Declare an explanatory variable, called x2, and assign it the result of dropping 'state' 'Name' and 'AdultWeekend' from the df2
x2 = df2.drop(['state','Name','AdultWeekend'], axis=1)
print(x2)

# Declare a response variable, called y2, and assign it the AdultWeekend column of the df2 
y2 = df2['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter x 
scaler2 = preprocessing.StandardScaler().fit(x2)

# Declare a variable called x2_scaled, and assign it the result of calling the transform() method with parameter x2 
x2_scaled=scaler2.transform(x2)

     summit_elev  vertical_drop  ...  NightSkiing_ac  clusters
0           3939           2500  ...           550.0         0
1           2600           1540  ...             0.0         0
2           2090            294  ...            30.0         0
3          11500           2300  ...             0.0         1
4          11100           1800  ...            80.0         1
..           ...            ...  ...             ...       ...
325         9500           1000  ...             0.0         1
326         7428            810  ...             0.0         1
327         7808           1571  ...           110.0         1
328         9663            990  ...             0.0         1
329         9500           1100  ...             0.0         1

[330 rows x 20 columns]


In [ ]:
# Get the 1-dimensional flattened array of our response variable y2 by calling the ravel() function on y2
y2 = y2.ravel()
# Call the train_test_split() function with the first two parameters set to x2_scaled and y2 
# Declare four variables, s2_train, s2_test, y2_train and y2_test separated by commas 
s2_train, s2_test, y2_train, y2_test = train_test_split(x2_scaled, y2, test_size=0.25, random_state=1)

In [ ]:
lm2 = linear_model.LinearRegression()
model2 = lm2.fit(s2_train,y2_train)

In [ ]:
# Make a variable called y2_pred and assign it the result of calling predict() on our model variable with parameter x2_test
y2_pred = model2.predict(s2_test)

In [ ]:
explained_variance_score(y_true=y2_test, y_pred=y2_pred)

0.93629622892495

In [ ]:
mean_absolute_error(y_true=y2_test, y_pred=y2_pred)

5.077794384195773

In [ ]:
lm2.intercept_

64.08028672751699

In [ ]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coef_data2 = []
coef_data2 = pd.DataFrame(abs(lm2.coef_), x2.columns, columns=['Coefficient'])
print(coef_data2)

                   Coefficient
summit_elev           4.591604
vertical_drop         0.959315
fastQuads             0.385538
quad                  0.806350
triple                1.072192
double                0.207273
surface               1.878713
total_chairs          2.029421
Runs                  1.673587
TerrainParks          0.043465
LongestRun_mi         0.385413
SkiableTerrain_ac     0.412156
Snow Making_ac        0.096407
daysOpenLastYear      0.694208
yearsOpen             0.281414
averageSnowfall       1.189317
AdultWeekday         19.890640
projectedDaysOpen     0.219737
NightSkiing_ac        0.758147
clusters              5.907850


In [ ]:
coef_data2.sort_values(['Coefficient'], ascending=False)

Coefficient
AdultWeekday         19.890640
clusters              5.907850
summit_elev           4.591604
total_chairs          2.029421
surface               1.878713
Runs                  1.673587
averageSnowfall       1.189317
triple                1.072192
vertical_drop         0.959315
quad                  0.806350
NightSkiing_ac        0.758147
daysOpenLastYear      0.694208
SkiableTerrain_ac     0.412156
fastQuads             0.385538
LongestRun_mi         0.385413
yearsOpen             0.281414
projectedDaysOpen     0.219737
double                0.207273
Snow Making_ac        0.096407
TerrainParks          0.043465

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

# **Model 3**

In [ ]:
df3 = pd.read_csv('step3_output.csv', index_col= None)

In [ ]:
df3.head()

Name    state  ...  NightSkiing_ac  clusters
0       Alyeska Resort   Alaska  ...           550.0         0
1  Eaglecrest Ski Area   Alaska  ...             0.0         0
2     Hilltop Ski Area   Alaska  ...            30.0         0
3     Arizona Snowbowl  Arizona  ...             0.0         1
4  Sunrise Park Resort  Arizona  ...            80.0         1

[5 rows x 23 columns]

In [ ]:
# Declare an explanatory variable, called x3, and assign it the result of dropping 'state', 'Name','summit_elev' and 'AdultWeekend' from the df3
x3 = df3.drop(['state','Name','AdultWeekend','summit_elev'], axis=1)
print(x3)

# Declare a response variable, called y3, and assign it the AdultWeekend column of the df3 
y3 = df3['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter x3 
scaler3 = preprocessing.StandardScaler().fit(x3)

# Declare a variable called x3_scaled, and assign it the result of calling the transform() method with parameter x3 
x3_scaled=scaler3.transform(x3)

     vertical_drop  fastQuads  ...  NightSkiing_ac  clusters
0             2500          2  ...           550.0         0
1             1540          0  ...             0.0         0
2              294          0  ...            30.0         0
3             2300          0  ...             0.0         1
4             1800          1  ...            80.0         1
..             ...        ...  ...             ...       ...
325           1000          0  ...             0.0         1
326            810          0  ...             0.0         1
327           1571          0  ...           110.0         1
328            990          0  ...             0.0         1
329           1100          0  ...             0.0         1

[330 rows x 19 columns]


In [ ]:
# Get the 1-dimensional flattened array of our response variable y3 by calling the ravel() function on y3
y3 = y3.ravel()
# Call the train_test_split() function with the first two parameters set to x3_scaled and y3 
# Declare four variables, s3_train, s3_test, y3_train and y3_test separated by commas 
s3_train, s3_test, y3_train, y3_test = train_test_split(x3_scaled, y3, test_size=0.25, random_state=1)

In [ ]:
lm3 = linear_model.LinearRegression()
model3 = lm3.fit(s3_train,y3_train)

In [ ]:
# Make a variable called y3_pred and assign it the result of calling predict() on our model variable with parameter x3_test
y3_pred = model3.predict(s3_test)

In [ ]:
explained_variance_score(y_true=y3_test, y_pred=y3_pred)

0.9364480858374188

In [ ]:
mean_absolute_error(y_true=y3_test, y_pred=y3_pred)

5.109325559805557

In [ ]:
lm3.intercept_

64.06901046671689

In [ ]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coef_data3 = []
coef_data3 = pd.DataFrame(abs(lm3.coef_), x3.columns, columns=['Coefficient'])
print(coef_data3)

                   Coefficient
vertical_drop         1.921695
fastQuads             0.137244
quad                  0.894367
triple                1.076065
double                0.163335
surface               1.730043
total_chairs          2.019026
Runs                  1.283589
TerrainParks          0.209494
LongestRun_mi         0.327486
SkiableTerrain_ac     0.136558
Snow Making_ac        0.048890
daysOpenLastYear      1.221096
yearsOpen             0.246508
averageSnowfall       0.967788
AdultWeekday         20.128488
projectedDaysOpen     0.569715
NightSkiing_ac        0.421403
clusters              2.630496


In [ ]:
coef_data3.sort_values(['Coefficient'], ascending=False)

Coefficient
AdultWeekday         20.128488
clusters              2.630496
total_chairs          2.019026
vertical_drop         1.921695
surface               1.730043
Runs                  1.283589
daysOpenLastYear      1.221096
triple                1.076065
averageSnowfall       0.967788
quad                  0.894367
projectedDaysOpen     0.569715
NightSkiing_ac        0.421403
LongestRun_mi         0.327486
yearsOpen             0.246508
TerrainParks          0.209494
double                0.163335
fastQuads             0.137244
SkiableTerrain_ac     0.136558
Snow Making_ac        0.048890

# **Model 4 - No Clusters Column**

In [ ]:
df4 = pd.read_csv('step3_output.csv', index_col= None)

In [ ]:
df4.head()

Name    state  ...  NightSkiing_ac  clusters
0       Alyeska Resort   Alaska  ...           550.0         0
1  Eaglecrest Ski Area   Alaska  ...             0.0         0
2     Hilltop Ski Area   Alaska  ...            30.0         0
3     Arizona Snowbowl  Arizona  ...             0.0         1
4  Sunrise Park Resort  Arizona  ...            80.0         1

[5 rows x 23 columns]

In [ ]:
# Declare an explanatory variable, called x4, and assign it the result of dropping 'clusters','state','Name','summit_elev' and 'AdultWeekend' from the df4
x4 = df4.drop(['state','Name','AdultWeekend','summit_elev','clusters'], axis=1)
print(x4)

# Declare a response variable, called y4, and assign it the AdultWeekend column of the df4 
y4 = df4['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter x4 
scaler4 = preprocessing.StandardScaler().fit(x4)

# Declare a variable called x4_scaled, and assign it the result of calling the transform() method with parameter x4 
x4_scaled=scaler4.transform(x4)

     vertical_drop  fastQuads  ...  projectedDaysOpen  NightSkiing_ac
0             2500          2  ...         150.000000           550.0
1             1540          0  ...          90.000000             0.0
2              294          0  ...         152.000000            30.0
3             2300          0  ...         122.000000             0.0
4             1800          1  ...         104.000000            80.0
..             ...        ...  ...                ...             ...
325           1000          0  ...         120.053004             0.0
326            810          0  ...          77.000000             0.0
327           1571          0  ...         123.000000           110.0
328            990          0  ...         120.053004             0.0
329           1100          0  ...         120.053004             0.0

[330 rows x 18 columns]


In [ ]:
# Get the 1-dimensional flattened array of our response variable y4 by calling the ravel() function on y4
y4 = y4.ravel()
# Call the train_test_split() function with the first two parameters set to x4_scaled and y4 
# Declare four variables, s4_train, s4_test, y4_train and y4_test separated by commas 
s4_train, s4_test, y4_train, y4_test = train_test_split(x4_scaled, y4, test_size=0.25, random_state=1)

In [ ]:
lm4 = linear_model.LinearRegression()
model4 = lm4.fit(s4_train,y4_train)

In [ ]:
# Make a variable called y4_pred and assign it the result of calling predict() on our model variable with parameter x4_test
y4_pred = model4.predict(s4_test)

In [ ]:
explained_variance_score(y_true=y4_test, y_pred=y4_pred)

0.9261715610757778

In [ ]:
mean_absolute_error(y_true=y4_test, y_pred=y4_pred)

5.464882310530267

In [ ]:
lm4.intercept_

64.04855122282349

In [ ]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coef_data4 = []
coef_data4 = pd.DataFrame(abs(lm4.coef_), x4.columns, columns=['Coefficient'])
print(coef_data4)

                   Coefficient
vertical_drop         1.029253
fastQuads             0.194817
quad                  1.257192
triple                0.982180
double                0.123181
surface               1.721121
total_chairs          2.162191
Runs                  1.172406
TerrainParks          0.124595
LongestRun_mi         0.207471
SkiableTerrain_ac     0.331341
Snow Making_ac        0.158166
daysOpenLastYear      1.047563
yearsOpen             0.300011
averageSnowfall       2.236228
AdultWeekday         20.138964
projectedDaysOpen     0.224700
NightSkiing_ac        0.522734


In [ ]:
coef_data4.sort_values(['Coefficient'], ascending=False)

Coefficient
AdultWeekday         20.138964
averageSnowfall       2.236228
total_chairs          2.162191
surface               1.721121
quad                  1.257192
Runs                  1.172406
daysOpenLastYear      1.047563
vertical_drop         1.029253
triple                0.982180
NightSkiing_ac        0.522734
SkiableTerrain_ac     0.331341
yearsOpen             0.300011
projectedDaysOpen     0.224700
LongestRun_mi         0.207471
fastQuads             0.194817
Snow Making_ac        0.158166
TerrainParks          0.124595
double                0.123181

# **Model 5: Drop AdultWeekday Column**

In [ ]:
df5 = pd.read_csv('step3_output.csv', index_col= None)

In [ ]:
df5.head()

Name    state  ...  NightSkiing_ac  clusters
0       Alyeska Resort   Alaska  ...           550.0         0
1  Eaglecrest Ski Area   Alaska  ...             0.0         0
2     Hilltop Ski Area   Alaska  ...            30.0         0
3     Arizona Snowbowl  Arizona  ...             0.0         1
4  Sunrise Park Resort  Arizona  ...            80.0         1

[5 rows x 23 columns]

In [ ]:
# Declare an explanatory variable, called x5, and assign it the result of dropping 'AdultWeekday','clusters','state','Name','summit_elev' and 'AdultWeekend' from the df5
x5 = df5.drop(['state','Name','AdultWeekend','summit_elev','clusters','AdultWeekday'], axis=1)
print(x5)

# Declare a response variable, called y5, and assign it the AdultWeekend column of the df5 
y5 = df5['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter x5 
scaler5 = preprocessing.StandardScaler().fit(x5)

# Declare a variable called x5_scaled, and assign it the result of calling the transform() method with parameter x5 
x5_scaled=scaler5.transform(x5)

     vertical_drop  fastQuads  ...  projectedDaysOpen  NightSkiing_ac
0             2500          2  ...         150.000000           550.0
1             1540          0  ...          90.000000             0.0
2              294          0  ...         152.000000            30.0
3             2300          0  ...         122.000000             0.0
4             1800          1  ...         104.000000            80.0
..             ...        ...  ...                ...             ...
325           1000          0  ...         120.053004             0.0
326            810          0  ...          77.000000             0.0
327           1571          0  ...         123.000000           110.0
328            990          0  ...         120.053004             0.0
329           1100          0  ...         120.053004             0.0

[330 rows x 17 columns]


In [ ]:
# Get the 1-dimensional flattened array of our response variable y5 by calling the ravel() function on y5
y5 = y5.ravel()
# Call the train_test_split() function with the first two parameters set to x5_scaled and y5 
# Declare four variables, s5_train, s5_test, y5_train and y5_test separated by commas 
s5_train, s5_test, y5_train, y5_test = train_test_split(x5_scaled, y5, test_size=0.25, random_state=1)

In [ ]:
lm5 = linear_model.LinearRegression()
model5 = lm5.fit(s5_train,y5_train)

In [ ]:
# Make a variable called y5_pred and assign it the result of calling predict() on our model variable with parameter x5_test
y5_pred = model5.predict(s5_test)

In [ ]:
explained_variance_score(y_true=y5_test, y_pred=y5_pred)

0.3643409369319489

In [ ]:
mean_absolute_error(y_true=y5_test, y_pred=y5_pred)

14.748422726744042

In [ ]:
lm5.intercept_

64.10316422131282

In [ ]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coef_data5 = []
coef_data5 = pd.DataFrame(abs(lm5.coef_), x5.columns, columns=['Coefficient'])
print(coef_data5)

                   Coefficient
vertical_drop         7.145961
fastQuads             9.658970
quad                  7.608741
triple                7.935227
double                6.023848
surface               2.819755
total_chairs         14.890137
Runs                  1.099856
TerrainParks          1.618422
LongestRun_mi         0.172467
SkiableTerrain_ac     0.997600
Snow Making_ac        5.924048
daysOpenLastYear      1.677280
yearsOpen             0.801812
averageSnowfall       1.091145
projectedDaysOpen     1.765574
NightSkiing_ac        0.646481


In [ ]:
coef_data5.sort_values(['Coefficient'], ascending=False)

Coefficient
total_chairs         14.890137
fastQuads             9.658970
triple                7.935227
quad                  7.608741
vertical_drop         7.145961
double                6.023848
Snow Making_ac        5.924048
surface               2.819755
projectedDaysOpen     1.765574
daysOpenLastYear      1.677280
TerrainParks          1.618422
Runs                  1.099856
averageSnowfall       1.091145
SkiableTerrain_ac     0.997600
yearsOpen             0.801812
NightSkiing_ac        0.646481
LongestRun_mi         0.172467

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.9417 |  4.85 |'base_elev'|
| Model 2. | 0.9363 |  5.08 |'state' base_elev'|
| Model 3. | 0.9364 |  5.11 |'state' 'summit_elev' 'base_elev'|
| Model 4. | 0.9262 |  5.46 |'state' 'summit_elev' 'base_elev' 'clusters'|
| Model 5. | 0.3643 | 14.75 |'state' 'summit_elev' 'base_elev' 'clusters' 'AdultWeekday'|

Model Selection: Model 1